pilot/pkg/serviceregistry/serviceentry/conversion.go

```diff
@@ -334,17 +340,20 @@ func getTLSModeFromWorkloadEntry(wle *networking.WorkloadEntry) string {
 
 // The foreign service instance has pointer to the foreign service and its service port.
 // We need to create our own but we can retain the endpoint already created.
-// TODO(rshriram): we currently ignore the pod(endpoint) ports and setup 1-1 mapping
-// from service port to endpoint port. Need to figure out a way to map k8s pod port to
-// appropriate service entry port
 func convertForeignServiceInstances(foreignInstance *model.ServiceInstance, serviceEntryServices []*model.Service,
 	serviceEntry *networking.ServiceEntry) []*model.ServiceInstance {
 	out := make([]*model.ServiceInstance, 0)
 	for _, service := range serviceEntryServices {
 		for _, serviceEntryPort := range serviceEntry.Ports {
 			ep := *foreignInstance.Endpoint
 			ep.ServicePortName = serviceEntryPort.Name
-			ep.EndpointPort = serviceEntryPort.Number
+			// if target port is set, use the target port else fallback to the service port
+			// TODO: we need a way to get the container port map from k8s
```

Comment:
hmm yeah this is not great right now since we are taking Endpoints as input which already have the targetport mapping done. We really need to change this to take in Pods directly (or info from a pod, at least) to fully solve

Reply:
Costin as against this from the very start. What we have here is the result of working with the existing abstractions. That said, I am not sure there are lots of people who want to do this ports by name. Esp with the mixed mode service, there is co-ordination required to ensure the named ports on workload entries match with those on k8s pods.